### ISSUE TITLE
[question] CONAN_DEFAULT_PACKAGE_ID_MODE=full_version_mode changes deps hash

### ISSUE THREAD

**Issue by boussaffawalid** [conan-io/conan#6855]
Hello,

I have a **package A** that depends on  a **package B** that depends on **KWidgetsAddons**
```
package A -> package B -> KWidgetsAddons
```

When I build my **package A** using:
```
CONAN_DEFAULT_PACKAGE_ID_MODE=semver_direct_mode
```
KWidgetsAddons package is being found in my local cache, and I get this in my build log

```
Build requirements packages
      KWidgetsAddons/5.54.0-gd140f3b@walid/stable:c3cc60322acd1d60cbcbe6d2842b6b9210358c8d - Cache
```
So far everything works as expected :) 
----

Now, when I change:
```
CONAN_DEFAULT_PACKAGE_ID_MODE=full_version_mode
```

KWidgetsAddons cannot be found anymore!

I got this is my log:
```
Build requirements packages
      KWidgetsAddons/5.54.0-gd140f3b@walid/stable:863882c3127f5efda5fc94cb5f4f5f22b3559431 - Missing
```

I have the same issue with few other dependencies as well but not all of them.

What is wrong here? and why my deps hash changes when I change the CONAN_DEFAULT_PACKAGE_ID_MODE ? 

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
I think this is totally expected. That is how the ``default_package_id_mode``. If you change the ``default_package_id_mode``, it will change all the package ids, because they are computed differently. The mode applies to all nodes in the graph, all the nodes will get a new package ID. Maybe only the header-only libraries that do not depend on others, and doesn't have settings/options, will keep the same package ID, but that would be the exception.

Does KWidgetsAddons have dependendencies?

**Comment by boussaffawalid** [comment]
KWidgetsAddons have one dependency (qt) which is also a dependency for Package A and B.

I don't really get this! as the issue appears for some dependencies but **NOT all of them**  

My understanding:
- if I `set  default_package_id_mode= semver_direct_mode` any package that matches `KWidgetsAddons/5.54.*` will be used.

- if I `set  default_package_id_mode= full_version_mode` any package that matches `KWidgetsAddons/5.54.0-gd140f3b` will be used.

In both scenarios `KWidgetsAddons/5.54.0-gd140f3b@walid/stable` should match.

Did I miss anything here?

**Comment by memsharded** [comment]
I think there is some misunderstanding of the ``package_id`` modes. They define how every package computes its ``package_id``, not how packages are consumed. It is not related to version-ranges matchings, but to the different binaries that are built.

> KWidgetsAddons have one dependency (qt) which is also a dependency for Package A and B.

Then, that is the reason. The KWidgetsAddons ``package_id`` computed with different modes is different, because in one case, it will hash ``qt/1.2.3`` (full_version), and in the other it will hash ``qt/1.X.Y`` (same for all X.Y, semver mode).

Please tell me if this clarifies it a bit.

**Comment by boussaffawalid** [comment]
now it make sense, thanks.

I have two questions here: 

If I built my deps with `default_package_id_mode=full_version_mode`, will I be able to use them upstream with both `full_version_mode` and `semver_direct_mode` ? 

Second question: I want to build everything with `semver_direct_mode` and only enforce full version match in one package (Package A), is there an easy way todo that ?

**Comment by memsharded** [comment]
> If I built my deps with default_package_id_mode=full_version_mode, will I be able to use them upstream with both full_version_mode and semver_direct_mode ?

No. the package_id_mode is something that should be defined and stick with it forever, and keep it consistent in all servers and clients wanting to use those packages.

> Second question: I want to build everything with semver_direct_mode and only enforce full version match in one package (Package A), is there an easy way todo that ?

Yes, this is definitely possible. Go to the package ``package_id()`` method:

```python
def package_id(self):
     self.info.requires["packageA"].full_version_mode()
```
Have a look to:
- https://docs.conan.io/en/latest/reference/conanfile/methods.html#package-id
- https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#dependency-issues

**Comment by boussaffawalid** [comment]
Thanks for the clarification, I was using two different default_package_id_mode and that was the root of all troubles.

**Comment by memsharded** [comment]
Lets add a clarification in the docs then, thanks :)

**Comment by boussaffawalid** [comment]
that would very helpful, thanks

**Comment by memsharded** [comment]
Closed by conan-io/docs#1666, thanks!
